### PR TITLE
Block Examples: Skip and warn if block type is missing

### PR DESCRIPTION
**Pull request by marekhrabe**
## Description
As reported in #19815, block previews could cause editor fatals if some blocks used for the preview are missing.

Easy way to trigger this is to unregister the Image block `wp.blocks.unregisterBlockType( 'core/image' )` and navigate to the block inserter where you hover Columns. Columns internally use Image for the example.

What I've tried in this PR is to check if block type exists before using it in Preview. If the block type is missing, a dev warning is shown and the block is skipped from the preview. This only applies to the example preview, nothing has been changed about missing block types in the post content and other places.

Since Examples are a visual enhancement, I thought it might be ok to just skip unknown blocks without any visual error or fatal. Also, the most common usage of Examples would only use the block itself (which would always be registered) so this situation is quite limited to blocks that use some other blocks as their example children.

## How has this been tested?
1. Open Editor
2. Run `wp.blocks.unregisterBlockType( 'core/image' )`
3. Open Inserter
4. Hover Columns block
5. On master, you should see Fatal Error. On this branch, you should see the example sans Images and a warning in console

## Screenshots 

| Columns Example | Columns with Image unregistered | 
| - | - |
| Screenshot 2020-02-06 at 14 30 46 | Screenshot 2020-02-06 at 14 30 55

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by youknowriad — COMMENTED** [review]
When there's no "example" at all, we should "Preview not available", I wonder if we should do the same in this case?

File: packages/blocks/src/api/factory.js [hunk 1]
```diff
@@ -537,11 +538,17 @@ export function switchToBlockType( blocks, name ) {
  * @return {Object} block.
  */
 export const getBlockFromExample = ( name, example ) => {
+	if ( ! getBlockType( name ) ) {
+		warning( 'Block example used an unknown block type: ' + name );
+		return false;
```
Comment (talldan) [inline packages/blocks/src/api/factory.js]: The return type of the function would need to be updated in the doc block to be something like `@return {Object|false} block.`

I personally feel `undefined` would be a better return value than `false` to indicate the absence of a block, but I realise this is a nitpick 😄. If the return value were `undefined`, the doc block would then be `@return {?Object} block.` (I think that's how they're still declared).

File: packages/blocks/src/api/factory.js [hunk 2]
```diff
@@ -537,11 +538,17 @@ export function switchToBlockType( blocks, name ) {
...
+	}
 	return createBlock(
 		name,
 		example.attributes,
-		map( example.innerBlocks, ( innerBlock ) =>
-			getBlockFromExample( innerBlock.name, innerBlock )
+		filter(
```
[the remark this answers is not in the document]
Comment (getdave) [reply packages/blocks/src/api/factory.js]: Minor: whilst I understand where Dan is coming from, "filter out falsey values" feels pretty easy to comprehend whereas `compact` requires some mental gymnastics to remember how it works (it's a Lodash API not a JS specific one so it presumes an awareness of Lodash).

[the remark this answers is not in the document]
Comment (talldan) [reply packages/blocks/src/api/factory.js]: Though you can't call the JS version without a function, so for me, I still had to look up the usage and test it in the console.

[the remark this answers is not in the document]
Comment (marekhrabe) [reply packages/blocks/src/api/factory.js]: I've switched to `compact`. I think both of these require a `lodash` knowledge and I'm not attached to one or the other 🤷‍♂️

**Comment by marekhrabe** [comment]
> When there's no "example" at all, we should "Preview not available", I wonder if we should do the same in this case?

@youknowriad I've looked at all usages of `getBlockFromExample` and one of them (packages/block-editor/src/components/inserter/menu.js), as you mention, tests for a presence of `.example` on a block and if it's not there, it shows "No Preview Available". 

Looking at the code, I think we don't need any extra checks with this change. When `getBlockFromExample` is called from all current usages, the call is always based on a block definition that's registered and the root block that `getBlockFromExample` creates is always the block itself, thus guaranteed to exist.

File: packages/blocks/src/api/factory.js (hunk 1 above)
Comment (marekhrabe) [reply packages/blocks/src/api/factory.js]: good idea with `undefined`. does the code/docs look better now?

File: packages/blocks/src/api/factory.js [hunk 3]
```diff
@@ -530,18 +532,25 @@ export function switchToBlockType( blocks, name ) {
 
 /**
  * Create a block object from the example API.
+ * Can return `undefined` if the block can't be created.
  *
  * @param {string} name
  * @param {Object} example
  *
- * @return {Object} block.
+ * @return {?Object} block object or undefined.
```
Comment (aduth) [inline packages/blocks/src/api/factory.js]: This syntax is for nullable types.

```suggestion
 * @return {Object|undefined} block object or undefined.
```

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#nullable-undefined-and-void-types

File: packages/blocks/src/api/factory.js [hunk 4]
```diff
@@ -530,18 +532,25 @@ export function switchToBlockType( blocks, name ) {
 
 /**
  * Create a block object from the example API.
+ * Can return `undefined` if the block can't be created.
```
Comment (aduth) [inline packages/blocks/src/api/factory.js]: Might be good to expand on when this could expect to happen.

**Comment by aduth** [comment]
> Looking at the code, I think we don't need any extra checks with this change. When `getBlockFromExample` is called from all current usages, the call is always based on a block definition that's registered and the root block that `getBlockFromExample` creates is always the block itself, thus guaranteed to exist.

I'm having a hard time interpreting this. Would that mean we will expect the "No Preview Available" with these changes? Or that it's not necessary? I guess based on the implementation, my worry would be that we allow some partially-functioning preview (a column with _some_ of its intended child examples), where personally I think a total bail-out might be the best approach if the integrity can't be guaranteed. Might not be as big an issue for known cases (columns), but potentially more of an issue for other blocks.

I don't hold too strong an opinion here, so I'm fine to be overruled.

**Comment by marekhrabe** [comment]
> I'm having a hard time interpreting this. Would that mean we will expect the "No Preview Available" with these changes? Or that it's not necessary? 

Sorry, it's a bit difficult to explain. I don't think it's necessary to add any additional check. From places where `getBlockFromExample` is currently called, it is always guaranteed that the block at the root of the preview is registered. 

Example with the inserter - you need to hover a block tile in order to show the example. You are only able to hover blocks that are registered at the moment. If you unregister a block, its tile will disappear from the inserter and you cannot hover it and trigger the example anymore. 

The only time a block could be missing is that if your example contains `innerBlocks` definition. 

> I guess based on the implementation, my worry would be that we allow some partially-functioning preview (a column with some of its intended child examples), where personally I think a total bail-out might be the best approach if the integrity can't be guaranteed. Might not be as big an issue for known cases (columns), but potentially more of an issue for other blocks.

Skipping the rendering (and warning) seemed like the easiest way to remove fatal errors so that's the approach I've tried first. Whatever path we choose, we should document the usage in docs https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#example-optional as I have just noticed it's completely missing a mention of `innerBlocks` at the moment. I can make it part of this PR once we agree on the approach

**Comment by youknowriad** [comment]
I have the same concerns as Andrew. I think if we find any missing block in the inner blocks, we should disable the entire preview. In simple blocks like columns and groups that is ok but imagine the "social link" block is missing when previewing "social links" block (or button when previewing buttons), it breaks the preview entirely.

**Comment by marekhrabe** [comment]
I can rework this to remove previews if any inner block type is missing. Would you prefer to do this as part of this PR or separately? It might need a bit of refactoring due to the way examples are called deeply nested inside render functions like here 

https://github.com/WordPress/gutenberg/blob/ce048553e6101571fee0ee8842ae18358b0a8162/packages/block-editor/src/components/block-switcher/index.js#L206-L229

and putting more checks inside it would probably be a bad idea from code readability perspective. I guess preview could be moved to the root of `render` or separated into a function.

My thinking was that potentially incomplete examples would already be an improvement from doing a full-screen fatal but I will defer the decision to you.

**Comment by talldan** [comment]
An ErrorBoundary might be another idea @marekhrabe. The idea being to let errors happen in the previews, but catch them with an ErrorBoundary and instead show the message about the preview being unavailable.

**Comment by aduth** [comment]
> An ErrorBoundary might be another idea @marekhrabe. The idea being to let errors happen in the previews, but catch them with an ErrorBoundary and instead show the message about the preview being unavailable.

I'd not love that there are errors in the first place, but it does seem like something we should have anyways, so it could serve as at least an interim solution to the problem 🤷‍♂

**Comment by paaljoachim** [comment]
What is needed to move this PR onward?
Can we get a status update?

Thanks!

**Comment by youknowriad** [comment]
I think this PR is not relevant anymore. Thanks for your efforts.

**Pull request closed without merging**